pkg/pillar/localcommand/agent.go

```diff
@@ -0,0 +1,564 @@
+	// StillRunning touches a file per agentName to signal the event loop is still running
+	// Those files are observed by the watchdog
+	StillRunning(agentName string, warnTime, errTime time.Duration)
+	// CheckMaxTimeTopic verifies if the time for a call has exceeded a reasonable number.
+	CheckMaxTimeTopic(agentName, topic string, start time.Time,
+		warnTime, errTime time.Duration)
+}
+
+// ConfigAgent applies commands and partial configuration changes received
+// locally from LPS or LOC.
+// This is implemented by zedagent (see pkg/pillar/cmd/zedagent/localcommand.go).
+type ConfigAgent interface {
+	// ApplyProfile applies a given profile, which then decides the subset of applications
+	// to activate.
+	ApplyProfile(string)
+
+	// ApplyRadioSilence applies radio silence configuration.
+	ApplyRadioSilence(types.RadioSilence)
+
+	// ApplyLocalDeviceCommand applies a locally received device command and reports
+	// if any changes were actually triggered.
+	ApplyLocalDeviceCommand(cmd types.DevCommand, timestamp uint64) (triggeredChanges bool)
+
+	// ApplyLocalAppRestartCmd applies a locally requested restart command for an app.
+	ApplyLocalAppRestartCmd(appUUID uuid.UUID, localCmd types.AppInstanceOpsCmd)
+
+	// ApplyLocalAppPurgeCmd applies a locally requested purge command for an app.
+	ApplyLocalAppPurgeCmd(appUUID uuid.UUID, localCmd types.AppInstanceOpsCmd,
+		localVolumeGenCounters map[string]int64)
+}
+
+// PubSubTopicReader : methods used by LocalCmdAgent to read messages from pubsub topics.
+// LocalCmdAgent only reads from pubsub topics, it does not change anything.
+// It is up to ConfigAgent to make and publish any changes.
+type PubSubTopicReader interface {
+	// GetAll returns all published messages.
+	GetAll() map[string]interface{}
+	// Get returns message identified by the given key.
+	Get(key string) (interface{}, error)
+}
+
+// lpsAddresses stores discovered LPS addresses grouped by network interface.
+// addrDiscoveryMx is used to serialize calls to discoverLpsAddresses().
+// LocalCmdAgent tasks that read from this map do not need to take the lock,
+// since they are paused while LPS address discovery is in progress.
+type lpsAddresses struct {
+	addrsByIface    map[string][]lpsAddress // key = interface name, value = addresses
+	upToDate        bool
+	addrDiscoveryMx sync.Mutex
+}
+
+// Returns true if there are no stored LPS addresses.
+func (la *lpsAddresses) empty() bool {
+	for _, addrs := range la.addrsByIface {
+		if len(addrs) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// taskTicker wraps a flextimer.FlexTickerHandle with synchronization
+// and a throttle flag. It allows dynamically adjusting ticker intervals,
+// forcing immediate ticks, and temporarily suppressing ticks when throttled.
+type taskTicker struct {
+	sync.RWMutex
+	tickerHandle flextimer.FlexTickerHandle
+	throttled    bool
+}
+
+// newTaskTicker creates a new taskTicker with a randomized firing interval.
```

Comment:
The research paper from the 1980's says you need 0.5 to 1.5 to avoid synchronization.